art/defences/trainer/certified_trainer_pytorch.py

```diff
@@ -0,0 +1,374 @@
+                  shape (nb_samples,).
+        :param certification_loss: Which certification loss function to use. Either "interval_loss_cce"
+                                   or "max_logit_loss". By default will use interval_loss_cce.
+                                   Alternatively, a user can supply their own loss function which takes in as input
+                                   the zonotope predictions of the form () and labels of the from () and returns a
+                                   scalar loss.
+        :param batch_size: Size of batches to use for certified training. NB, this will run the data
+                           sequentially accumulating gradients over the batch size.
+        :param nb_epochs: Number of epochs to use for training.
+        :param training_mode: `True` for model set to training mode and `'False` for model set to evaluation mode.
+        :param scheduler: Learning rate scheduler to run at the start of every epoch.
+        :param kwargs: Dictionary of framework-specific arguments. This parameter is not currently supported for PyTorch
+               and providing it takes no effect.
+        """
+        import torch  # lgtm [py/repeated-import]
+
+        if batch_size is None:
+            batch_size = self.batch_size
+
+        if nb_epochs is not None:
+            epochs: int = nb_epochs
+        elif self.batch_size is not None:
+            epochs = self.batch_size
+
+        # Set model mode
+        self._classifier._model.train(mode=training_mode)  # pylint: disable=W0212
+
+        if self._classifier._optimizer is None:  # pragma: no cover # pylint: disable=W0212
+            raise ValueError("An optimizer is needed to train the model, but none is provided.")
+
+        y = check_and_transform_label_format(y, nb_classes=self._classifier.nb_classes)
+
+        # Apply preprocessing
+        x_preprocessed, y_preprocessed = self._classifier.apply_preprocessing(x, y, fit=True)
+
+        # Check label shape
+        y_preprocessed = self._classifier.reduce_labels(y_preprocessed)
+
+        num_batch = int(np.ceil(len(x_preprocessed) / float(self.pgd_params["batch_size"])))
+        ind = np.arange(len(x_preprocessed))
+        from sklearn.utils import shuffle
+
+        x_cert = np.copy(x_preprocessed)
+        y_cert = np.copy(y_preprocessed)
+
+        # Start training
+        if self.use_certification_schedule:
+            if self.certification_schedule is None:
+                certification_schedule_function = DefaultLinearScheduler(
+                    step_per_epoch=self.bound / epochs, initial_bound=0.0
+                )
+        else:
+            bound = self.bound
+
+        for _ in tqdm(range(epochs)):
+            if self.use_certification_schedule:
+                bound = certification_schedule_function.step()
+            # Shuffle the examples
+            random.shuffle(ind)
+
+            # Train for one epoch
+            pbar = tqdm(range(num_batch))
```

Comment:
How about a verbose argument for the progress bar?

Reply:
Added the option to toggle verbosity